src/hotspot/share/gc/g1/heapRegion.cpp

```diff
@@ -459,153 +459,71 @@ void HeapRegion::print_on(outputStream* st) const {
-  oop _containing_obj;
-  bool _failures;
-  int _n_failures;
-  VerifyOption _vo;
-
-public:
-
-  G1VerificationClosure(G1CollectedHeap* g1h, VerifyOption vo) :
-    _g1h(g1h), _ct(g1h->card_table()),
-    _containing_obj(NULL), _failures(false), _n_failures(0), _vo(vo) {
-  }
-
-  void set_containing_obj(oop obj) {
-    _containing_obj = obj;
-  }
-
-  bool failures() { return _failures; }
-  int n_failures() { return _n_failures; }
-
-  void print_object(outputStream* out, oop obj) {
-#ifdef PRODUCT
-    Klass* k = obj->klass();
-    const char* class_name = k->external_name();
-    out->print_cr("class name %s", class_name);
-#else // PRODUCT
-    obj->print_on(out);
-#endif // PRODUCT
-  }
-};
-
-class VerifyLiveClosure : public G1VerificationClosure {
 
   template <class T>
   void do_oop_work(T* p) {
-    assert(_containing_obj != NULL, "Precondition");
-    assert(!_g1h->is_obj_dead_cond(_containing_obj, _vo),
-      "Precondition");
+    assert(_containing_obj != nullptr, "Precondition");
+    assert(!_g1h->is_obj_dead_cond(_containing_obj, _vo), "Precondition");
 
     T heap_oop = RawAccess<>::oop_load(p);
     if (CompressedOops::is_null(heap_oop)) {
       return;
     }
 
     oop obj = CompressedOops::decode_raw_not_null(heap_oop);
-    bool is_in_heap = _g1h->is_in(obj);
-    if (!is_in_heap || _g1h->is_obj_dead_cond(obj, _vo)) {
-      MutexLocker x(ParGCRareEvent_lock, Mutex::_no_safepoint_check_flag);
 
-      Log(gc, verify) log;
-      if (!_failures) {
-        log.error("----------");
-      }
-      ResourceMark rm;
+    HeapRegion* from = _g1h->heap_region_containing_or_null(p);
+    HeapRegion* to = _g1h->heap_region_containing_or_null(obj);
+    if (from != nullptr && to != nullptr &&
+      from != to &&
+      !to->is_pinned() &&
+      to->rem_set()->is_complete()) {
 
-      HeapRegion* from = _g1h->heap_region_containing(p);
-      log.error("Field " PTR_FORMAT " of live obj " PTR_FORMAT " in region " HR_FORMAT,
-                p2i(p), p2i(_containing_obj), HR_FORMAT_PARAMS(from));
-      LogStream ls(log.error());
-      print_object(&ls, _containing_obj);
+      jbyte cv_obj = *_ct->byte_for_const(_containing_obj);
+      jbyte cv_field = *_ct->byte_for_const(p);
+      const jbyte dirty = G1CardTable::dirty_card_val();
 
-      if (!is_in_heap) {
-        log.error("points to address " PTR_FORMAT " outside of heap", p2i(obj));
-      } else {
-        HeapRegion* to = _g1h->heap_region_containing(obj);
-        log.error("points to dead obj " PTR_FORMAT " in region " HR_FORMAT " remset %s",
+      bool is_bad = !(from->is_young()
+        || to->rem_set()->contains_reference(p)
+        || (_containing_obj->is_objArray() ?
+              cv_field == dirty :
+              cv_obj == dirty || cv_field == dirty));
+
+      if (is_bad) {
+        ResourceMark rm;
+        Log(gc, verify) log;
+        LogStream ls(log.error());
+
+        MutexLocker x(ParGCRareEvent_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
[pre-existing] That's a weird name for a mutex that seems to only be used by G1.

Reply:
Historically it had been used in CMS and G1 (and Serial, but I think it just got victimized here because of shared code), but since CMS is dead now... I'll rename it in a separate CR.